### PR TITLE
Add file logger callback & export train loss json file

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
This PR adds a new callback, which is invoked on the logging event - if the most recent log contains specific keys, the subdictionary is extracted, and the `train_loss.json` is reexported.

Example using LLama 7b with the twitter complaints example (tested on single GPU with a v100).

```bash
python tuning/sft_trainer.py  --model_name_or_path $MODEL_PATH  --data_path $DATA_PATH  --output_dir $OUTPUT_PATH  --num_train_epochs 5  --per_device_train_batch_size 4  --per_device_eval_batch_size 4  --gradient_accumulation_steps 4  --evaluation_strategy "no"  --save_strategy "epoch"  --learning_rate 1e-5  --weight_decay 0.  --warmup_ratio 0.03  --lr_scheduler_type "cosine"  --logging_steps 1  --include_tokens_per_second  --packing False  --response_template "\n### Label:"  --peft_method pt --tokenizer_name_or_path $MODEL_PATH --dataset_text_field "output" --torch_dtype "float32" --use_flash_attn False
```

Produces a `train_loss.json` file in the output directory:
```bash
$ ls out
checkpoint-13  checkpoint-15  checkpoint-3  checkpoint-6  checkpoint-9  train_loss.jsonl
```

Here is an example of the json file that gets created.
```json
{"data": {"epoch": 0.31, "step": 1, "timestamp": "2024-02-13T12:29:27.387128", "value": 10.6876}, "name": "loss"}
{"data": {"epoch": 0.62, "step": 2, "timestamp": "2024-02-13T12:29:28.078879", "value": 10.8365}, "name": "loss"}
{"data": {"epoch": 0.92, "step": 3, "timestamp": "2024-02-13T12:29:28.749249", "value": 10.767}, "name": "loss"}
{"data": {"epoch": 1.23, "step": 4, "timestamp": "2024-02-13T12:29:31.796401", "value": 10.8091}, "name": "loss"}
{"data": {"epoch": 1.54, "step": 5, "timestamp": "2024-02-13T12:29:32.436218", "value": 10.7143}, "name": "loss"}
{"data": {"epoch": 1.85, "step": 6, "timestamp": "2024-02-13T12:29:33.187178", "value": 10.7687}, "name": "loss"}
{"data": {"epoch": 2.15, "step": 7, "timestamp": "2024-02-13T12:29:36.343676", "value": 10.7223}, "name": "loss"}
{"data": {"epoch": 2.46, "step": 8, "timestamp": "2024-02-13T12:29:37.021809", "value": 10.7292}, "name": "loss"}
{"data": {"epoch": 2.77, "step": 9, "timestamp": "2024-02-13T12:29:37.708132", "value": 10.8619}, "name": "loss"}
{"data": {"epoch": 3.08, "step": 10, "timestamp": "2024-02-13T12:29:42.506561", "value": 10.6114}, "name": "loss"}
{"data": {"epoch": 3.38, "step": 11, "timestamp": "2024-02-13T12:29:43.154734", "value": 10.6282}, "name": "loss"}
{"data": {"epoch": 3.69, "step": 12, "timestamp": "2024-02-13T12:29:43.838959", "value": 10.7648}, "name": "loss"}
{"data": {"epoch": 4.0, "step": 13, "timestamp": "2024-02-13T12:29:44.538090", "value": 10.8187}, "name": "loss"}
{"data": {"epoch": 4.31, "step": 14, "timestamp": "2024-02-13T12:29:48.925495", "value": 10.6256}, "name": "loss"}
{"data": {"epoch": 4.62, "step": 15, "timestamp": "2024-02-13T12:29:49.682787", "value": 10.7308}, "name": "loss"}
```

Note that logs are only exported from the main process to avoid writing from mulitple processes for multigpu tunings.

### CODE REVIEW

File: tuning/sft_trainer.py [hunk 1]
```diff
@@ -29,6 +31,40 @@ def on_save(self, args, state, control, **kwargs):
...
+                # Take the subdict of the last log line; if any log_keys aren't part of this log
+                # object, asssume this line is something else, e.g., train completion, and skip.
+                log_obj = {k: logs[k] for k in FileLoggingCallback.log_keys}
+            except KeyError:
+                return
+
+            # Redump the json file in the checkpoint directory with the updated log line
+            self.existing_logs.append(log_obj)
+            with open(log_file_path, "w") as log_file:
+                json.dump(self.existing_logs, log_file, sort_keys=True, indent=4)
```
Comment (gkumbhat) [inline tuning/sft_trainer.py:61]: One thing I would be a bit worried about is the need to open and close file repeatedly. Also looks like we are overriding the existing logs instead of appending, any particular reason for that?

File: tuning/sft_trainer.py [hunk 2]
```diff
@@ -29,6 +31,40 @@ def on_save(self, args, state, control, **kwargs):
...
+    """Exports metrics, e.g., training loss to a file in the checkpoint directory."""
+    # Keys to be exported from the log dict
+    log_keys = ["loss", "epoch"]
+
+    def __init__(self, logger):
+        self.logger = logger
+        self.existing_logs = []
+
+    def on_log(self, args, state, control, logs=None, **kwargs):
+        """Checks if this log contains keys of interest, e.g., los, and if so, creates
```
Comment (gkumbhat) [inline tuning/sft_trainer.py:44]:
```suggestion
        """Checks if this log contains keys of interest, e.g., loss, and if so, creates
```

File: tuning/sft_trainer.py [hunk 3]
```diff
@@ -29,6 +31,40 @@ def on_save(self, args, state, control, **kwargs):
...
+        train_loss.json in the model output dir (if it doesn't already exist),
+        appends the subdict of the log & dumps the file.
+        """
+
+        log_file_path = os.path.join(args.output_dir, "train_loss.json")
```
Comment (gkumbhat) [inline tuning/sft_trainer.py:49]: Can we add either a conversion or a check to make sure `loss` is actually in `float` instead of other data types. Otherwise the json dump would fail.

**Review by gkumbhat — COMMENTED** [review]
- General question, should we follow the same pattern as used by `caikit-nlp` for consistency and compatibility perspective ?
- We'll need to see how the callbacks work for steps vs epochs. Since we may want to log individual steps, even if it is set to epoch wise decisions. Another scenario we'll need to understand better is multi-gpu case.

File: tuning/sft_trainer.py (hunk 1 above)
Comment (alex-jw-brooks) [reply tuning/sft_trainer.py:61]: Just that this is writing a JSON file, so we need to reparse it to add to the list. We could alternatively write a list of JSON objects (i.e., each line is a JSON object representing one log), and in that case just append per log?

I agree that it's a bit of a bummer to open and close the file so much, although it's still probably quite small compared to the actual training of the model, and the parsing is the main problem. If we write a json object per log, we can keep it open in append mode and flush on each written log I guess?

**Comment by alex-jw-brooks** [comment]
Hey @gkumbhat! Thanks for the review. Some thoughts on your broader questions

> General question, should we follow the same pattern as used by caikit-nlp for consistency and compatibility perspective ?

I would greatly prefer to avoid subclassing the trainer just for logging. The actual logging events look similar either way, so any logic in the logging should be pretty portable between the two approaches, and this project is already using callbacks, so callbacks feels more natural. Plus the SFT trainer is already a trainer subclass and pretty complex, subclassing it again just to write an extra file seems like unnecessary complexity to me.

> We'll need to see how the callbacks work for steps vs epochs. Since we may want to log individual steps, even if it is set to epoch wise decisions. Another scenario we'll need to understand better is multi-gpu case.
For steps vs epochs etc - the event being handled here just gives a parsed version of the logged object, so we can just handle those keys separately, I think. I'll look into making sure this behavior is sensible.

For multigpu, we probably need to make sure that the file is only written by the master process, but I'm not sure how rank is taken into account for each process. I can look into this as well, although do you know if we formally support multiGPU at the moment, or is it still experimental?

**Comment by gkumbhat** [comment]
> I would greatly prefer to avoid subclassing the trainer just for logging. The actual logging events look similar either way, so any logic in the logging should be pretty portable between the two approaches, and this project is already using callbacks, so callbacks feels more natural. Plus the SFT trainer is already a trainer subclass and pretty complex, subclassing it again just to write an extra file seems like unnecessary complexity to me, since it would basically be for the one function

But follow pattern, I actually meant the formatting of the output file. Sorry I was not quite clear there

**Comment by gkumbhat** [comment]
> do you know if we formally support multiGPU at the moment, or is it still experimental?
Ops, I don't know 🤷

**Comment by alex-jw-brooks** [comment]
Thanks everyone - updated the PR to write in jsonl format, and only dump logs from process zero to prevent clobbering when running multiprocess training

**Comment by alex-jw-brooks** [comment]
Hey @gkumbhat - updated the PR and description based on our discussions to match the legacy log format, should be ready for another look when you have a moment

**Review by gkumbhat — APPROVED** [review]
